src/emu_seproxyhal.c

```diff
@@ -12,27 +12,26 @@ static bool status_sent = false;
 static uint8_t last_tag;
 static size_t next_length;
 
-static ssize_t read_helper(int fd, void *buf, size_t count)
+static ssize_t readall(int fd, void *buf, size_t count)
 {
   ssize_t n;
 
-  while (1) {
+  while (count > 0) {
     /* actually issue a syscall which is forwarded to the host */
     n = read(fd, buf, count);
-    if (n > 0) {
-      break;
-    } else if (n == 0) {
+    if (n == 0) {
       warnx("read from seph fd failed: fd closed");
       return -1;
-    } else {
+    } else if (n < 0) {
       if (errno == EINTR)
         continue;
       warn("read from seph fd failed");
       return -1;
     }
+    count -= n;
   }
 
-  return n;
+  return count;
```

Comment:
Returning `count`, which is decremented, is strange: now `readall` returns the number of bytes which were *not* read. And later in `sys_io_seproxyhal_spi_recv` the returned size is not really used, which is normal (as the received size is the size given in `readall` call).

Maybe change the result as "error status" and use `return 0` in `readall` ?